src/keys/public.rs

```diff
@@ -0,0 +1,109 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use dusk_schnorr::NotePublicKey;
+
+use crate::{permutation, StealthAddress};
+
+use dusk_jubjub::{JubJubAffine, JubJubExtended, JubJubScalar};
+
+use super::secret::SecretKey;
+
+#[cfg(feature = "rkyv-impl")]
+use rkyv::{Archive, Deserialize, Serialize};
+
+use dusk_bytes::{DeserializableSlice, Error, HexDebug, Serializable};
+use dusk_jubjub::GENERATOR_EXTENDED;
+use subtle::{Choice, ConstantTimeEq};
+
+/// Public pair of `a·G` and `b·G` defining a [`PublicKey`]
+#[derive(HexDebug, Clone, Copy)]
+#[cfg_attr(
+    feature = "rkyv-impl",
+    derive(Archive, Serialize, Deserialize),
+    archive_attr(derive(bytecheck::CheckBytes))
+)]
+pub struct PublicKey {
+    A: JubJubExtended,
+    B: JubJubExtended,
+}
+
+impl PublicKey {
+    /// This method is used to construct a new `PublicKey` from the given
+    /// public pair of `a·G` and `b·G`
+    pub fn new(A: JubJubExtended, B: JubJubExtended) -> Self {
+        Self { A, B }
+    }
+
+    /// Gets `A` (`a·G`)
+    pub fn A(&self) -> &JubJubExtended {
+        &self.A
+    }
+
+    /// Gets `B` (`b·G`)
+    pub fn B(&self) -> &JubJubExtended {
+        &self.B
+    }
+
+    /// Generates new `PKr = H(A · r) · G + B` from a given `r`
+    pub fn gen_stealth_address(&self, r: &JubJubScalar) -> StealthAddress {
+        let G = GENERATOR_EXTENDED;
+        let R = G * r;
+
+        let rA = self.A * r;
+        let rA = permutation::hash(&rA);
+        let rA = G * rA;
+
+        let pk_r = rA + self.B;
+        let pk_r = NotePublicKey::from(pk_r);
```

Comment:
I'm not sure if I still like this even though I suggested it earlier.

Maybe this actually also works and looks a bit cleaner
```suggestion
        let pk_r: NotePublicKey = pk_r.into();
```